server/jukebox.py

```diff
@@ -0,0 +1,94 @@
+from server.client_manager import ClientManager
+from dataclasses import dataclass
+from collections import Counter
+from typing import Union, Tuple
+from threading import Timer
+import random
+import asyncio
+@dataclass
+class JukeboxVote:
+    """Represents a single vote cast for the jukebox."""
+    client: ClientManager.Client
+    name: str
+    length: int
+    showname: str
+    chance: int = 1
+
+class Jukebox:
+    def __init__(self, area):
+        self.area = area
+        self.votes = []
+        self.jukebox_prev_char_id = -1
+        self.music_looper = None
+        self.playing = False
+
+    def add_jukebox_vote(self, client: ClientManager.Client, music_name: str, length: int = -1, showname: str = ''):
+
+        # If a length for a song doesn't exist, the server
+        # doesn't know when to play the next song
+        if length == -1:
+            client.send_ooc(f'{music_name} needs a length set.')
+            return
+
+        # You can only have one vote 
+        # in the queue at a time
+        self.remove_jukebox_vote(client)
+
+        self.votes.append(JukeboxVote(client, music_name, length, showname))
+        client.send_ooc('Your song was added to the jukebox.')
+
+        if len(self.votes) > 0:
+            self.play()
+
+    def play(self):
+        song_picked, song_length = self.get_next_song_in_queue()
+        if song_picked == '':
+            self.area.broadcast_ooc('No other songs in the voting queue.')
+            return 
+
+        if self.playing is False:
+            self.area.send_command('MC', song_picked, song_length)
+            self.playing = True
+            timer_for_next_song = Timer(song_length, self.set_not_playing)
```

Comment:
use `asyncio.call_later`. also how does it play the next track if the timer is always hooked to `set_not_playing`?

Reply:
It sets the jukebox to `not_playing`. `set_not_playing` sets `playing` to `False` and then runs `Play` again. Then empties out the jukebox queue for the next song

Reply:
I think I described that horribly. But I did use `asyncio.call_later`